go/libraries/doltcore/doltdb/commit_hooks.go

```diff
@@ -127,3 +184,69 @@ func (lh *LogHook) SetLogger(ctx context.Context, wr io.Writer) error {
 	lh.out = wr
 	return nil
 }
+
+func RunAsyncReplicationThreads(ctx context.Context, wg *sync.WaitGroup, ch chan PushArg, destDB *DoltDB, tmpDir string, logger io.Writer) {
+	mu := &sync.Mutex{}
+	var newHeads = make(map[string]PushArg, asyncPushBufferSize)
+
+	// stop first go routine before second
+	newCtx, stop := context.WithCancel(context.Background())
+	wg.Add(1)
+	go func() error {
+		defer wg.Done()
+		for {
+			select {
+			case p, ok := <-ch:
+				if !ok {
+					return ctx.Err()
+				}
+				mu.Lock()
+				newHeads[p.ds.ID()] = p
+				mu.Unlock()
+			case <-ctx.Done():
+				stop()
+				return ctx.Err()
+			}
+		}
+	}()
+
+	wg.Add(1)
+	go func() error {
+		defer wg.Done()
+		defer close(ch)
+		var latestHeads = make(map[string]hash.Hash, asyncPushBufferSize)
+		var newHeadsCopy = make(map[string]PushArg, asyncPushBufferSize)
+		flush := func() {
+			mu.Lock()
+			if len(newHeads) == 0 {
+				mu.Unlock()
+				return
+			}
+			for k, v := range newHeads {
+				newHeadsCopy[k] = v
+			}
+			mu.Unlock()
```

Comment:
I think I would move this logic into a function that can use a defer block to unlock. something like:

```
getHeadsCopy() map[string]PushArg {
	mu.Lock()
	defer mu.Unlock()
	if len(newHeads) == 0 {
		return nil
	}
		
	var newHeadsCopy = make(map[string]PushArg, asyncPushBufferSize)
	for k, v := range newHeads {
		newHeadsCopy[k] = v
	}
	return newHeadsCopy
}
```

Reply:
This also clears out the newHeadsCopy each loop... which seems like what you would want.